R/fb_filter_species_traits.R

```diff
@@ -0,0 +1,258 @@
+#' 
+#' @description
+#' For each trait computes the percentage of species without `NA` (missing 
+#' trait values).
+#' 
+#' @inheritParams fb_get_coverage
+#'
+#' @return A three-column `data.frame` with:
+#' - `trait`: the name of the trait;
+#' - `n_species`: the number of species with non-missing value for the trait;
+#' - `coverage`: the percentage of species with non-missing value for the trait.
+#' 
+#' @export
+#'
+#' @examples
+#' library("funbiogeo")
+#' 
+#' data("species_traits")
+#' 
+#' species_coverage_by_trait <- fb_count_species_by_traits(species_traits)
+
+fb_count_species_by_traits <- function(species_traits) {
+  
+  ## Check inputs ----
+  
+  if (missing(species_traits)) {
+    stop("Argument 'species_traits' (species x traits data frame) is required")
+  }
+  
+  check_species_traits(species_traits)
+  
+  
+  ## Compute species coverage by trait ----
+  
+  species_coverage <- unlist(lapply(colnames(species_traits)[-1], function(x) 
+    sum(!is.na(species_traits[[x]]))))
+  
+  species_coverage <- data.frame(
+    "trait"      = colnames(species_traits)[-1],
+    "n_species"  = species_coverage,
+    "coverage"   = species_coverage / nrow(species_traits))
+  
+  species_coverage <- species_coverage[order(species_coverage$"coverage", 
+                                             decreasing = TRUE), ]
+  
+  rownames(species_coverage) <- NULL
+  
+  species_coverage
+}
+
+
+
+#' Filter traits with a given species coverage threshold
+#' 
+#' @description
+#' Selects traits (columns) for which the percentage of species without 
+#' `NA` (missing trait values) is higher than a threshold.
+#' 
+#' @param threshold a numeric of length 1 between 0 and 1. The percentage of 
+#' species coverage threshold.
+#' 
+#' @inheritParams fb_get_coverage
+#'
+#' @return A subset of `species_traits` with traits covered by X% of species.
+#' 
+#' @export
+#'
+#' @examples
+#' library("funbiogeo")
+#' 
+#' data("species_traits")
+#' 
+#' species_traits <- fb_filter_traits_by_species_coverage(species_traits,
+#'                                                        threshold = 0.6)
+
+fb_filter_traits_by_species_coverage <- function(species_traits, 
+                                                 threshold = 0) {
+  
+  ## Check inputs ----
+  
+  if (missing(species_traits)) {
+    stop("Argument 'species_traits' (species x traits data frame) is required")
+  }
+  
+  check_species_traits(species_traits)
+  
+  if (!is.numeric(threshold) | threshold > 1 | threshold < 0) {
+    stop("Coverage threshold should be a numeric value >= 0 and <= 1",
+         call. = FALSE)
+  }
+  
+  
+  # Check for absence of variability in traits ----
+  
+  n_modalities  <- unlist(lapply(colnames(species_traits)[-1], function(x) 
+    length(unique(species_traits[[x]][!is.na(species_traits[[x]])]))))
+  
+  only_na_traits <- colnames(species_traits)[-1][which(n_modalities == 0)]
+  
+  if (length(only_na_traits)) {
```

Comment:
Length equal to what? What happens if length = numeric(0) (same argument for all`if (length()){}` afterwards)

Reply:
Done in 3735192